WIP: Benchmarking

SRFU-NN opened the pull request:
# BenchmarkInstance
A class representing a single test. Minimal input is
* model_system_name: str, The system to test. Must be built in at the moment. 
* xpyrimentor_definition: dict, The dict that defines an XPyrimentor to test. Allows for setting most properties.
* experimental_budget: int, When to discard the optimisation as "failed"
* expected_random_runtime: float, How hard the optimisation problem should be
* seed: int, Ensures reproducibility and different behavior among replicates.

# NoiseModel
You can now set arbitrary noise distributions by e.g.
```python
    noise_model.noise_types["constant"] = lambda: 2
    noise_model.noise_type = "constant"
```
In this instance it is used to get points two standard deviations above the mean value, but it is generally applicable.

SRFU-NN committed:
chore: Added flake8 settings

SRFU-NN committed:
chore: Allow to set seed from ModelSystem initialisor

SRFU-NN committed:
chore: NoiseModel now uses the fully fledged RNG getter

SRFU-NN committed:
doc: Started on benchmarking notebook

SRFU-NN committed:
feat: Moved benchmarking function to separate file

SRFU-NN committed:
Doc: Benchmarking

SRFU-NN committed:
feat: Seed for LHS

SRFU-NN committed:
doc: Seeding LHS in benchmarking

SRFU-NN committed:
chore: Change noise type to property function

SRFU-NN committed:
chore: Changed named lambda functions to def functions

SRFU-NN committed:
fix: Changed noise model to noise type being a property

SRFU-NN committed:
feat: ModelSystem now has a copy method

SRFU-NN committed:
doc: Finding limits in benchmarking

SRFU-NN committed:
feat: Better naming for benchmarking notebook

SRFU-NN committed:
Merge remote-tracking branch 'origin/develop' into benchmarking

SRFU-NN committed:
doc: Notes on benchmarking

SRFU-NN committed:
feat: Started benchamrking script

SRFU-NN committed:
feat: Golden ratio sampling suggestor

SRFU-NN committed:
Merge branch 'golden_ratio_sampling' into benchmarking

SRFU-NN committed:
feat: Benchamrking on a single instance

SRFU-NN committed:
feat: Benchmarking with validation

SRFU-NN committed:
chore: Restored defaults

SRFU-NN committed:
doc: Reference on Generalized Golden Ratio Sequence

SRFU-NN committed:
feat: ConstantSuggestor

SRFU-NN committed:
fix: Integer.sample([1]) gave a point above the high limit

SRFU-NN committed:
chore: Robusting hart3 against non-float inputs

SRFU-NN committed:
feat: Benchmarking framework done

SRFU-NN committed:
doc: More benchmanrking options

SRFU-NN committed:
chore: Removed temporary script

SRFU-NN commented:
Noise bounds - Allow setting noise bounds in the Optimizer initialiser. As kwargs
Also length scale bounds as kwargs.

Look at #95 , it is about this.

SRFU-NN commented:
Remove white noise - Add white noise while estimating success, remove it afterwards.

Edit: Done

SRFU-NN committed:
Merge remote-tracking branch 'origin/develop' into benchmarking

SRFU-NN committed:
refactor: Noise distributions is now a dict to allow for easy addition

SRFU-NN committed:
chore: Benchmark uses new way to set noise distribution

SRFU-NN committed:
chore: Benchmark uses new way to set noise distribution

SRFU-NN committed:
doc: Updated XPyriMentor notebook

SRFU-NN committed:
doc: Saving Suggestor definition in benchamrikng notebook

SRFU-NN committed:
Merge branch 'benchmarking' of https://github.com/novonordisk-research/ProcessOptimizer into benchmarking

SRFU-NN committed:
doc: Writing why NoiseModel.noise_types is a dict with lambda functions as values

SRFU-NN committed:
chore: NoiseModel.normal convenience function

SRFU-NN committed:
Merge remote-tracking branch 'origin/develop' into benchmarking

SRFU-NN committed:
chore: Moved benchmarking

SRFU-NN committed:
refactor: Made uniform noise into class method

Lambdas can't be pickled, so having noise as lambdas made running benchmarks in parallel impossible

SRFU-NN committed:
fix: Make benchmarkInstance importable after move

SRFU-NN committed:
Merge branch 'develop' into benchmarking

SRFU-NN committed:
Merge branch 'benchmarking' of https://github.com/novonordisk-research/ProcessOptimizer into benchmarking

SRFU-NN commented:
> Noise bounds - Allow setting noise bounds in the Optimizer initialiser. As kwargs Also length scale bounds as kwargs.
> 
> Look at #95 , it is about this.

Length scale bounds now settable

SRFU-NN committed:
feat: Use observational noise in benchmarking

SRFU-NN committed:
doc: Notebooks

SRFU-NN committed:
feat: Length scale bounds work in Benchmarking

SRFU-NN committed:
Merge remote-tracking branch 'origin/develop' into benchmarking

SRFU-NN committed:
feat: Exposing noise level bounds in Optimizer

SRFU-NN committed:
feat: Noise bounds forwarded to GPR

SRFU-NN committed:
doc: Updated benchamrking

RuneChristensen-NN commented on ProcessOptimizer/XpyriMentor/suggestors/constant_suggestor.py:
```diff
@@ -0,0 +1,37 @@
+        if not is_listlike(point):
+            point = [point] * space.n_dims
+        if convert:
+            point = space.sample(point)
```
So "sample" is in this case really just transforming from unit space into the space provided in the Space object. Is that a sample?

SRFU-NN replied:
ConstantSuggestor shouldn't sample, it should require a point in real space. Implemented in #351

RuneChristensen-NN commented on ProcessOptimizer/XpyriMentor/suggestors/constant_suggestor.py:
```diff
@@ -0,0 +1,37 @@
+        self.point = np.array(point)
+
+    def suggest(
+            self, Xi: Iterable[Iterable], Yi: Iterable, n_asked: int = 1
```
Not clear what the purpose of Xi and Yi is?

RuneChristensen-NN commented on ProcessOptimizer/XpyriMentor/xpyrimentor.py:
```diff
@@ -71,6 +71,7 @@ def ask(self, n: int = 1) -> np.ndarray:
         Ask the suggestor for new points to evaluate. The number of points to ask is
         specified by the argument n. The method returns a list of new points to evaluate.
         """
+        n = int(n) # Ensure that n is an integer
```
What happens if you cannot make the conversion to int?

RuneChristensen-NN commented on examples/features/control_parameters.ipynb:
```diff
@@ -601,7 +601,7 @@
    "name": "python",
    "nbconvert_exporter": "python",
    "pygments_lexer": "ipython3",
-   "version": "3.11.5"
```
Is this change intentional?

SRFU-NN replied:
No, just something that happens with jupyter notebooks when you run them.

RuneChristensen-NN commented on ProcessOptimizer/XpyriMentor/suggestors/golden_ratio_suggestor.py:
```diff
@@ -0,0 +1,42 @@
+        self.offset = self.rng.random()
+
+    @staticmethod
+    def phi(d):
```
It appears d must be 1 or larger. It will also have to be an integer for it to me meaningful. I see have it comes naturally in the suggest method, but being explicit about that could be an option. Right now we are solving brute force numerically for 10 iterations in all cases and arriving at approximate solutions. For lower dimensions, we should be able to find exact solutions and use those instead.

RuneChristensen-NN commented on ProcessOptimizer/XpyriMentor/suggestors/golden_ratio_suggestor.py:
```diff
@@ -0,0 +1,42 @@
+            x = pow(1+x,1/(d+1)) 
+        return x
+
+    def suggest(
```
Where is Yi sued? Maybe a bit of comments would make this more readable

RuneChristensen-NN commented on ProcessOptimizer/XpyriMentor/suggestors/suggestor_factory.py:
```diff
@@ -46,6 +48,7 @@ def suggestor_factory(
         logger.debug("Creating DefaultSuggestor")
         return DefaultSuggestor(space, n_objectives, rng)
     try:
+        definition = definition.copy()
```
Would like a comment on why this is needed

RuneChristensen-NN commented on ProcessOptimizer/XpyriMentor/suggestors/suggestor_factory.py:
```diff
@@ -96,6 +99,7 @@ def suggestor_factory(
         logger.debug("Creating SequentialStrategizer")
         suggestors = []
         for suggestor in definition["suggestors"]:
+            suggestor: dict = suggestor.copy()
```
Same: would like a comment on why you need to do the .copy()

RuneChristensen-NN commented on ProcessOptimizer/XpyriMentor/suggestors/suggestor_factory.py:
```diff
@@ -111,5 +115,15 @@ def suggestor_factory(
                 suggestor_factory(space, suggestor, n_objectives, rng, n_points = n)
             ))
         return SequentialStrategizer(suggestors)
+    elif suggestor_type == "GoldenRatio":
```
The series of "elif" works but it could get rather long over time if more suggestors are added. Could we find a more general way of doing it. Maybe a dict that stores suggestor_type keys and corresponding suggestors with kwargs?

SRFU-NN replied:
We could move interpreting the definition to a class method in each suggestor. Though the strategizers would need to call the factory, which could give us issues with circular references. I'll look at it.

SRFU-NN replied:
https://github.com/novonordisk-research/ProcessOptimizer/issues/349

RuneChristensen-NN commented on ProcessOptimizer/benchmarking/benchmark.py:
```diff
@@ -0,0 +1,154 @@
+    # value.
+    model_system.noise_model.noise_types["constant"] = lambda: 2
+    model_system.noise_model.noise_type = "constant"
+    return model_system.get_score(x)
```
Blank line at end?

RuneChristensen-NN commented on ProcessOptimizer/benchmarking/benchmark.py:
```diff
@@ -0,0 +1,154 @@
+            How "hard" the system is to optimize. This is the expected number of random
+            parameter sets you need to evaluate to find a good one.
+        * `seed` [int]:
+            Random seed to use.
```
optional variables are not described. It would also be great if the purpose of having this class was stated somewhere.

RuneChristensen-NN commented on ProcessOptimizer/model_systems/hart3.py:
```diff
@@ -30,6 +30,10 @@ def hart3_score(x):
         The score of the system at x.
     """
     # Define the constants that are canonically used with this function.
+    if isinstance(x, np.ndarray):
```
If we start typechecking and converting to arrays of dtype float, why stop there?
Maybe making a try-except -> raise error at the x = np.asarray(x, dtype=float) step?

SRFU-NN replied:
I can't remember why I did this, it seems to work without it.

I had some issues with dtype, where numpy really likes to make everything text if even one input is text. But all suggestors explicitly return it is dtype=object, which shouldn't give problems. I've removed the code.

RuneChristensen-NN commented on ProcessOptimizer/model_systems/hart3.py:
```diff
@@ -40,13 +44,14 @@ def hart3_score(x):
     return -np.sum(alpha * np.exp(-np.sum(A * (np.array(x) - P) ** 2, axis=1)))
```
The np.array(x) seems redundant now. On the other side, I guess the harm is negligible.

RuneChristensen-NN commented on ProcessOptimizer/model_systems/hart6.py:
```diff
@@ -31,6 +31,10 @@ def hart6_score(x):
         The score of the system at x.
     """
     # Define the constants that are canonically used with this function.
+    if isinstance(x, np.ndarray):
```
If we start typechecking and converting to arrays of dtype float, why stop there?
Maybe making a try-except -> raise error at the x = np.asarray(x, dtype=float) step?

RuneChristensen-NN commented on ProcessOptimizer/model_systems/hart6.py:
```diff
@@ -51,12 +55,13 @@ def hart6_score(x):
     return -np.sum(alpha * np.exp(-np.sum(A * (np.array(x) - P) ** 2, axis=1)))
```
The np.array(x) seems redundant now. On the other side, I guess the harm is negligible.

RuneChristensen-NN commented on ProcessOptimizer/space/space.py:
```diff
@@ -456,8 +456,9 @@ def distance(self, a, b):
     def _sample(self, point_list: Iterable[float]) -> np.ndarray:
         point_list = [
-            point * (self.high + 1 - self.low) + self.low for point in point_list
+            min(point * (self.high + 1 - self.low) + self.low,self.high) for point in point_list
```
Can you write a comment on why this is needed.
Also, is it just a transform we are doing here or what is going on?

SRFU-NN replied:
Commented in #351

_sample is explained in the base class Dimension, line 235. Is that enough, or should we explain it in all subclasses?

RuneChristensen-NN commented on ProcessOptimizer/tests/test_noise_model.py:
Great to see the lambda functions go

RuneChristensen-NN commented on ProcessOptimizer/tests/test_suggestors/test_constant_suggestor.py:
```diff
@@ -0,0 +1,41 @@
+    for suggestion in suggestions:
+        assert suggestion in space
+        assert len(suggestion) == 2
+        assert (suggestion == suggestions[0]).all()
```
Missing empty line at end?

RuneChristensen-NN commented on examples/design_qualifications/benchmarking.ipynb:
This looks like it still has some way to go. 
In any case, better descriptions of or comments in the code would be nice.

RuneChristensen-NN commented on examples/design_qualifications/golden_ratio_sampling.ipynb:
```diff
@@ -0,0 +1,61 @@
+    "import matplotlib.pyplot as plt\n",
+    "from XpyriMentor import XpyriMentor\n",
+    "\n",
+    "# Create a new instance of the XpyriMentor class\n",
```
Focus on that we are using XpyriMentor to load the GoldenRatio suggester

RuneChristensen-NN commented on examples/design_qualifications/golden_ratio_sampling.ipynb:
```diff
@@ -0,0 +1,61 @@
+    "from XpyriMentor import XpyriMentor\n",
+    "\n",
+    "# Create a new instance of the XpyriMentor class\n",
+    "xpyrimentor = XpyriMentor(space = [(0.0, 1.0) for _ in range(2)], suggestor={\"suggestor_name\": \"GoldenRatio\"})\n",
```
Maybe be more explicit with the definition of the space. For range(2) we are not saving anything by looping.

RuneChristensen-NN commented on examples/design_qualifications/golden_ratio_sampling.ipynb:
```diff
@@ -0,0 +1,61 @@
+    "\n",
+    "# Create a new instance of the XpyriMentor class\n",
+    "xpyrimentor = XpyriMentor(space = [(0.0, 1.0) for _ in range(2)], suggestor={\"suggestor_name\": \"GoldenRatio\"})\n",
+    "points = xpyrimentor.ask(150)\n",
```
Comment that we ask the GoldenRatio suggestor for 150 points.
It could be nice to show what happens for different number of points and whether you can pull more points after the initial ask

RuneChristensen-NN commented on ProcessOptimizer/model_systems/model_system.py:
```diff
@@ -11,42 +11,72 @@ class ModelSystem:
+        * `space` [List or Space]:
+            A list of dimension defintions or the parameter space as a Space object.
+
+        * `noise_model` [str, dict, or NoiseModel]:
```
Nice that you can use various distributions for the noise. Could that be made obvious here?

SRFU-NN replied:
I've written it in a notebook. At the moment, you have to go through the `NoiseModel`. If we want to write about it here, we should probably make it a method of `ModelSystem`, and then we can document it there.

RuneChristensen-NN commented on ProcessOptimizer/model_systems/noise_models.py:
```diff
@@ -32,17 +34,30 @@ def __init__(
         # Note that this has the potential for problems if _noise_distribution does not
-        # have "size" 1, but as long as it is only set by set_noise_type(), it should be
-        # safe.
+        # have "size" 1, but as long as it is only use the ones defined here, you should
```
grammar

RuneChristensen-NN commented on ProcessOptimizer/model_systems/noise_models.py:
```diff
@@ -32,17 +34,30 @@ def __init__(
-        # Change this to ..utils.get_random_genertor once the pull request with that have been merged
-        self.set_noise_type("normal")
+        self._rng = get_random_generator(seed)
+        self.noise_types = {
```
NICE!

RuneChristensen-NN commented on ProcessOptimizer/benchmarking/benchmark.py:
```diff
@@ -0,0 +1,154 @@
+        optimizer.remove_observational_noise()
+        return (result_location, result_value + 2*result_std)
+
+def run_benchmark(benchmark_instance: BenchmarkInstance) -> BenchmarkInstance:
```
We are not testing batch strategies. Only one function evaluation at a time

RuneChristensen-NN commented on ProcessOptimizer/benchmarking/benchmark.py:
I have not done a full review of this. I think it would be nice to have the code explained.

RuneChristensen-NN reviewed (commented):
Really nice. There are a couple of different things in here that could maybe have been two different PR, one with the suggestors and XpyriMentor and one with the benchmarking.

SRFU-NN committed:
doc: Results of code review

SRFU-NN committed:
Merge remote-tracking branch 'origin/develop' into benchmarking

SRFU-NN committed:
doc: Explaining the rough way to calculate phi_d

SRFU-NN committed:
fix: Only passing noise level bounds to Gaussian Process Regressor

SRFU-NN committed:
refactor: Run as method for benchmark instances

SRFU-NN committed:
Merge branch 'active_task' into benchmarking

SRFU-NN committed:
Merge branch 'suggestors' into benchmarking

SRFU-NN committed:
chore: Clean up import

SRFU-NN committed:
fix: Proteting suggestors from n_points added by sequntial_suggestor

SRFU-NN committed:
fix: XPyrimentor converts X's to numpy arrays

This allows uniform handling of e.g. lists

SRFU-NN committed:
fix: Copying suggestor definition

This preserves keywords that are popped in the original definition

SRFU-NN committed:
feat: benchmark.tell()

To allow for uniform handling and saving of intermediate estimates of the optimum

SRFU-NN committed:
feat: Benchmark report method

SRFU-NN committed:
chore: Minor updates in benchmark

SRFU-NN committed:
feat: All constant suggestors replicate numbers are reported from benchmarking

SRFU-NN committed:
fix: Stupid mistake

SRFU-NN committed:
doc: Reporitng on benchmarking

SRFU-NN committed:
test: Fixing tests for XPyrimentor converting x to nump arrays

SRFU-NN committed:
Merge remote-tracking branch 'origin/develop' into benchmarking

SRFU-NN committed:
chore: Renamed variable "xpyrimentor defintion"

It is just the suggestor definition

SRFU-NN committed:
doc: Docstring for BenchmarkInstance, describing the parameters

SRFU-NN committed:
chore: Removed redundant property BenchmarkInstance.model

Now only BenchmarkInstance.model_system exists

SRFU-NN committed:
doc: Explaining the benchmark.tell intricacies

SRFU-NN committed:
chore: Removed unnecessary modification

SRFU-NN committed:
doc: Specifying expectations of a Suggestor

SRFU-NN committed:
Merge remote-tracking branch 'origin/develop' into benchmarking

SRFU-NN committed:
test, fix: Corrected changed keyword

SRFU-NN committed:
test, fix: Doubled equal sign

SRFU-NN committed:
Merge branch 'develop' into benchmarking

SRFU-NN committed:
fix: Golden ratio suggestor offset away from diagonal

The generalized golden atio suggestor had an offset at a random place on the diagonal of its Space. Now it has an offest at a random place in its Space.

SRFU-NN committed:
fix: Hartmann 3 and Hartmann 6 model systems accept points from Xpyrimentor

Xpyrimentor suggest points as ndarrays with dtype object. Hartmann systems require is to be numerical, so we convert the input to floats before mathing

SRFU-NN committed:
chore: return type hint for Space.rvs()

SRFU-NN committed:
fix: Changed a variable name in benchmark

SRFU-NN committed:
doc: Changed example benchmark settings

SRFU-NN committed:
Merge branch 'benchmarking' of https://github.com/novonordisk-research/ProcessOptimizer into benchmarking

SRFU-NN committed:
chore: Typing, Iterable -> Collection where len is used.

SRFU-NN committed:
chore: Code review suggestions
